horovod/spark/torch/datamodule.py

```diff
@@ -0,0 +1,229 @@
+                 random_seed=0,
+                 **kwargs):
+        from petastorm import TransformSpec, make_reader, make_batch_reader
+
+        super().__init__(**kwargs)
+        self.reader_pool_type = reader_pool_type
+        self.train_reader_worker_count = train_reader_worker_count
+        self.val_reader_worker_count = val_reader_worker_count
+        self.random_seed = random_seed
+
+        # In general, make_batch_reader is faster than make_reader for reading the dataset.
+        # However, we found out that make_reader performs data transformations much faster than
+        # make_batch_reader with parallel worker processes. Therefore, the default reader
+        # we choose is make_batch_reader unless there are data transformations.
+        self.transform_spec = TransformSpec(self.transform_fn) if self.transform_fn else None
+        self.reader_factory_kwargs = dict()
+        if self.transform_spec:
+            self.reader_factory = make_reader
+            self.reader_factory_kwargs['pyarrow_serialize'] = True
+        else:
+            self.reader_factory = make_batch_reader
+
+    def __enter__(self):
+        super().__enter__()
+        # Petastorm: read data from the store with the correct shard for this rank
+        # setting num_epochs=None will cause an infinite iterator
+        # and enables ranks to perform training and validation with
+        # unequal number of samples
+        self.train_reader = self.reader_factory(
+            self.train_dir,
+            num_epochs=None,
+            cur_shard=hvd.rank(),
+            reader_pool_type=self.reader_pool_type,
+            workers_count=self.train_reader_worker_count,
+            shard_count=hvd.size(),
+            hdfs_driver=constants.PETASTORM_HDFS_DRIVER,
+            schema_fields=self.schema_fields,
+            transform_spec=self.transform_spec,
+            storage_options=self.storage_options,
+            shuffle_rows=self.shuffle,
+            shuffle_row_groups=self.shuffle,
+            seed=self.random_seed,
+            **self.reader_factory_kwargs
+        )
+
+        self.val_reader = self.reader_factory(
+            self.val_dir,
+            num_epochs=None,
+            cur_shard=hvd.rank(),
+            reader_pool_type=self.reader_pool_type,
+            workers_count=self.val_reader_worker_count,
+            shard_count=hvd.size(),
+            hdfs_driver=constants.PETASTORM_HDFS_DRIVER,
+            schema_fields=self.schema_fields,
+            transform_spec=self.transform_spec,
+            storage_options=self.storage_options,
+            shuffle_rows=False,
+            shuffle_row_groups=False,
+            seed=self.random_seed,
+            **self.reader_factory_kwargs
+        ) if self.has_val else self.empty_batch_reader()
+
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self.has_val and self.val_reader:
+            self.val_reader.__exit__(type, value, traceback)
```

Comment:
similar to above keras part, why `reader.stop()` is not used

Reply:
same as above.